src/gwt/acesupport/acemode/sql_highlight_rules.js

```diff
@@ -0,0 +1,94 @@
+/*
+* sql_highlight_rules.js
+*
+* Copyright (C) 2015 by RStudio, Inc.
+*
+* The Initial Developer of the Original Code is Jeffrey Arnold
+* Portions created by the Initial Developer are Copyright (C) 2014
+* the Initial Developer. All Rights Reserved.
+*
+* Unless you have received this program directly from RStudio pursuant
+* to the terms of a commercial license agreement with RStudio, then
+* this program is licensed to you under the terms of version 3 of the
+* GNU Affero General Public License. This program is distributed WITHOUT
+* ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+* MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+* AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+*
+*/
+
+define("mode/sql_highlight_rules", ["require", "exports", "module"], function(require, exports, module) {
+
+  var oop = require("ace/lib/oop");
+  var lang = require("ace/lib/lang");
+  var TextHighlightRules = require("ace/mode/text_highlight_rules").TextHighlightRules;
+
+  var SqlHighlightRules = function() {
+    var keywords = (
+        "select|insert|update|delete|from|where|and|or|group|by|order|limit|offset|having|as|case|" +
+        "when|else|end|type|left|right|join|on|outer|desc|asc|union|create|table|primary|key|if|" +
+        "foreign|not|references|default|null|inner|cross|natural|database|drop|grant"
+    );
+
+    var builtinConstants = (
+        "true|false"
+    );
+
+    var builtinFunctions = (
+        "avg|count|first|last|max|min|sum|ucase|lcase|mid|len|round|rank|now|format|" + 
+        "coalesce|ifnull|isnull|nvl"
+    );
+
+    var dataTypes = (
+        "int|numeric|decimal|date|varchar|char|bigint|float|double|bit|binary|text|set|timestamp|" +
+        "money|real|number|integer"
+    );
+
+    var keywordMapper = this.createKeywordMapper({
+        "support.function": builtinFunctions,
+        "keyword": keywords,
+        "constant.language": builtinConstants,
+        "storage.type": dataTypes
+    }, "identifier", true);
+
+    this.$rules = {
```

Comment:
We should do some special highlighting for the ? variable placeholder. I'd suggest one of the following:

"markup.list" (the pinkish color we use for delineating mathjax blocks; or
"comment.doc.tag" (light blue color we use for doxygen keywords)

Those are just two suggestions, you'll have to play with various alternatives to see what feels right.